packages/opentelemetry-context-async-hooks/src/AsyncLocalStorageContextManager.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Context } from '@opentelemetry/context-base';
+import { AsyncLocalStorage } from 'async_hooks';
+import { AbstractAsyncHooksContextManager } from './AbstractAsyncHooksContextManager';
+
+export class AsyncLocalStorageContextManager extends AbstractAsyncHooksContextManager {
+  private _asyncLocalStorage: AsyncLocalStorage<Context>;
+
+  constructor() {
+    super();
+    this._asyncLocalStorage = new AsyncLocalStorage();
```

Comment:
I think we should create the `_asyncLocalStorage` at the `init` and not in the constructor.

Reply:
I also though about this, but this essentially means we would need to guard every method for the case that `this._asyncLocalStorage` is `undefined` or `null`. Would there be any other benefit? 

Looking at the implementation of `AsyncLocalStorage` [itself](https://github.com/nodejs/node/blob/13c5a1629cd025ba560f34f6d3190b2f38d184d4/lib/async_hooks.js#L230), it is also only lazy enabled on the first `enterWith` [call](https://github.com/nodejs/node/blob/13c5a1629cd025ba560f34f6d3190b2f38d184d4/lib/async_hooks.js#L254), which is called by `ContentManager.with`

Reply:
I didn't know that it was lazy init'ed, which feel weird since async_hooks is not. Anyway thats fine for me in this case